provisioning/transport/mqtt/src/mqtt.ts

```diff
@@ -4,12 +4,413 @@
+      this._fsm.handle('disconnect', err, () => { });
+    };
+
+    this._mqttBase.on('message', responseHandler);
+    this._mqttBase.on('error', errorHandler);
+
+    this._fsm = new machina.Fsm({
+      namespace: 'provisioning-transport-mqtt',
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          _onEnter: (err, callback) => {
+            if (callback) {
+              callback(err);
+            } else if (err) {
+              this.emit('error', err);
+            }
+          },
+          registrationRequest: (request, callback) => {
+            /* Codes_SRS_NODE_PROVISIONING_MQTT_18_002: [ If the transport is not connected, `registrationRequest` shall connect it and subscribe to the response topic.] */
+            this._fsm.handle('connect', request, (err) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.transition('sendingRegistrationRequest', request, callback);
+              }
+            });
+          },
+          queryOperationStatus: (request, operationId, callback) => {
+            /* Codes_SRS_NODE_PROVISIONING_MQTT_18_016: [ If the transport is not connected, `queryOperationStatus` shall connect it and subscribe to the response topic.] */
+            this._fsm.handle('connect', request, (err) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.transition('sendingOperationStatusQuery', request, operationId, callback);
+              }
+            });
+          },
+          connect: (request, callback) => this._fsm.transition('connecting', request, callback),
+          /* Codes_SRS_NODE_PROVISIONING_MQTT_18_030: [ If `cancel` is called while the transport is disconnected, nothing will be done.] */
+          disconnect: (err, callback) => callback(err),
+          cancel: (callback) => callback()
+        },
+        connecting: {
+          _onEnter: (request, callback) => {
+            this._connect(request, (err) => {
+              if (err) {
+                this._fsm.handle('disconnect', err, callback);
+              } else {
+                this._fsm.transition('connected', request, null, callback);
+              }
+            });
+          },
+          disconnect: (err, callback) => this._fsm.transition('disconnecting', err, callback),
+          /* Codes_SRS_NODE_PROVISIONING_MQTT_18_031: [ If `cancel` is called while the transport is in the process of connecting, it shall wait until the connection is complete and then disconnect the transport.] */
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        connected: {
+          _onEnter: (request, result, callback) => {
+            if (this._deferredCancelInQueue()) {
+              debug('connect operation cancelled');
+              callback(new errors.OperationCancelledError());
+            } else {
+              callback(null, result);
+            }
```

Comment:
whoa took me a moment to understand that but that is pretty cool. now i'm wondering if we cancel correctly everywhere else.

Reply:
Unit tests FTW, but this is hacky -- looking ahead feels so wrong.  I'm wondering if we need a "cancelPending" flag that we set instead of deferring the cancel operation.

---
In reply to: [156274497](https://github.com/Azure/azure-iot-sdk-node/pull/178#discussion_r156274497) [](ancestors = 156274497)